### DISCUSSION TITLE
Collapse/expand all details boxes

### DISCUSSION THREAD

**Discussion by Hrovatin** [cotes2020/jekyll-theme-chirpy#2109] [Ideas]
### Description

To hide some "optional" content a details box can be used as follows, which can be opened on demand:

<details closed>
<summary><u>Title</u></summary>
<div style="border:1px solid #535353;border-radius: 7px;padding: 10px;margin: 5px;">
Details
</div>
</details>

by default, one wants to have details hidden to reduce amount of content. However, I have the habit of exporting blogs to PDFs so that I can underline and add comments when reading. In this case closed boxes are not good as one looses info. So Ideally one could easily expand all boxes before printing. This can be done with the following button, added directly to markdown of the post.

```
<button id="toggleDetails" style="border-radius: 7px; box-shadow: none; padding: 10px;">Expand/Collapse details</button>

<script>
document.addEventListener("DOMContentLoaded", function() {
    var detailsElements = document.querySelectorAll("details");
    var toggleButton = document.getElementById("toggleDetails");

    toggleButton.addEventListener("click", function() {
        var allOpen = Array.from(detailsElements).every(details => details.open);
        detailsElements.forEach(function(details) {
            details.open = !allOpen;
        });
    });
});
</script>
```

If anyone knows how to incorporate this in chirpy and make it more elegant, to reduce repetition, that would be very appreciated.

**Comment by tis-starlight** [comment]
This can be achieved in several ways:

#### 1. Make the `<details>` element open by default, allowing the user to collapse it by clicking.
```html
<details open>
  <summary>Expand/Collapse details</summary>
  <p>some text</p>
</details>
```

#### 2. Include the `<button>` element and the `toggle-details.js` script in the post where you want the toggle functionality to be enabled.

- Create a `toggle-details.js` script containing the code and place it under `/assets/js/toggle-details.js` :
```js
document.addEventListener('DOMContentLoaded', function () {
  var detailsElements = document.querySelectorAll('details');
  var toggleButton = document.getElementById('toggleDetails');

  toggleButton.addEventListener('click', function () {
    var allOpen = Array.from(detailsElements).every((details) => details.open);
    detailsElements.forEach(function (details) {
      details.open = !allOpen;
    });
  });
});
```
- Add the `<button>` element and script to the post:
```md
<!-- some post content -->

<button id="toggleDetails" style="border-radius: 7px; box-shadow: none; padding: 10px;">Expand/Collapse details</button>
<script src="/assets/js/toggle-details.js" defer></script>
```

#### 3. **Global implementation**
- Create the `toggle-details.js` script as mentioned in the previous method.
- Create a `metadata-hook.html` file (if you don’t have one already) and place it under `/_includes/metadata-hook.html`.
- Add the `toggle-details.js` script to the `metadata-hook.html` :
```html
<script src="/assets/js/toggle-details.js" defer></script>
```
With this, you no longer need to add the script to each post. You only need to include the `<button>` element to enable the toggle functionality.

<hr>

Additionally, If you export blogs to PDFs, you should consider looking into [Obsidian](https://obsidian.md) and [Pandoc](https://pandoc.org).

**Reply by Hrovatin** [reply]
Thank you very much. I will try it in the next days.

**Comment by Hrovatin** [comment]
I tried 2&3, but for some reason the script was not properly incorporated. Thus I used the below solution. I also improved the button appearance.

Make a file `_includes/toggle-details.html` with the below contents:

```
<style>
.toggle_switch {
  position: relative;
  display: inline-block;
  width: 60px;
  height: 34px;
  margin-left: 15px;
}

.toggle_switch input {
  opacity: 0;
  width: 0;
  height: 0;
}

.toggle_slider {
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: #ccc;
  transition: 0.4s;
  border-radius: 34px;
  border: 1px solid #999;
}

.toggle_slider:before {
  position: absolute;
  content: "";
  height: 26px;
  width: 26px;
  left: 4px;
  bottom: 4px;
  background-color: white;
  transition: 0.4s;
  border-radius: 50%;
}

input:checked + .toggle_slider {
  background-color: var(--toc-highlight);
}

input:checked + .toggle_slider:before {
  transform: translateX(26px);
}
</style>

<div style="margin-bottom: 20px;">
<span style="font-size: 0.9em; color: var(--text-muted-color);">
  Expand all detail boxes for printing
</span>
<label class="toggle_switch">
<input type="checkbox" id="toggleDetails">
<span class="toggle_slider"></span>
</label>
</div>

<script>
  document.addEventListener("DOMContentLoaded", function() {
    var detailsElements = document.querySelectorAll("details");
    var toggleSwitch = document.getElementById("toggleDetails");

    toggleSwitch.addEventListener("change", function() {
        var isChecked = toggleSwitch.checked;
        detailsElements.forEach(function(details) {
            details.open = isChecked;
        });
    });
  });
</script>

```

Where the button should be added in the post use:
```
{% include toggle-details.html %}
```
